libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -300,6 +300,15 @@ void WaveTrack::Interval::ClearAndAddCutLine(double t0, double t1)
       [&](auto& clip) { clip.ClearAndAddCutLine(t0, t1); });
 }
 
+void WaveTrack::Interval::AddCutLine(Interval &&interval)
+{
+   assert(NChannels() == interval.NChannels());
+   size_t ii = 0;
+   ForEachClip([&](auto &clip) {
+      clip.AddCutLine(std::move(interval).DetachClip(ii++));
```

Comment:
No.

`interval` is not a pointer.  It's a structure.

See how I defined `DetachClip` with the `&&` qualifier.  That forces `move` else it won't compile.

I defined it that way because the function steals data from the interval.

Reply:
@Paul-Licameli I received an email notification that you responded to that, but somehow the comment went lost. It went like this:

> No.
>
> interval is not a pointer. It's a structure.
>
> See how I defined DetachClip with the && qualifier. That forces move else it won't compile.
>
> I defined it that way because the function steals data from the interval.

I saw the declaration of `DetachClip` and can understand, yes, that the compiler will only accept it called on a right reference. But if the `move` has no side effect on a struct, then I don't see a reason for that signature.

Also, I made the following experiment
```
   struct Foo
   {
      int x = 1;
   };
   const auto foo = std::make_shared<Foo>();
   const auto bar = std::move(foo);
   REQUIRE(foo->x == 1);
   ```
which passes, i.e., the `Foo` instance has two strong references, so it doesn't look that `DetachClip` invalidates the member clip.

But @crsib is more knowledgeable than me in those things and if he does see a reason to do it that way then he can overtake this part of the review and I'd approve it as a whole. It may well be faster than you patiently explaining me what it is that I don't understand :D

Reply:
Your experiment applied move to the pointer but not to the object pointed at.

Reply:
Yes, but so does `DetachClip`:
```
WaveClipHolder DetachClip(size_t iChannel) &&
      { return move(iChannel == 0 ? mpClip : mpClip1); }
 ```
`mpClip` and `mpClip1` being shared pointers.

Reply:
> Yes, but so does `DetachClip`:
> 
> ```
> WaveClipHolder DetachClip(size_t iChannel) &&
>       { return move(iChannel == 0 ? mpClip : mpClip1); }
> ```
> 
> `mpClip` and `mpClip1` being shared pointers.

So this function has `&&` where you might otherwise use `const`.  It is a memebee function that must be called in an rvalue or moved object.

Reply:
Because `const` would be wrong.  `this` has a side-effect.  I could omit `&&` but putting it there forces `this` to be an rvalue or be moved.

Reply:
> Because `const` would be wrong. `this` has a side-effect. I could omit `&&` but putting it there forces `this` to be an rvalue or be moved.

Apologies for the late reply, somehow I missed this response and thought this work was on hold.
In my `Foo` example above, `foo` doesn't get invalidated by the `std::move(foo)` call apparently because it is `const`. I tried again without `const`, and indeed `move` then invalidated it.
`Interval::mpClip` and `mpClip1` are also `const`, though. I went as far as building and hitting that `DetachClip` implementation with a breakpoint: the `move` doesn't have the side effect you expect.
So I stand by opinion that this unusual method signature isn't justified.

Reply:
Aha, you are right